*** ../qmail-1.03.orig/qmail-smtpd.c	Mon Jun 15 19:53:16 1998
--- qmail-smtpd.c	Mon Feb 14 23:06:14 2000
***************
*** 24,29 ****
--- 24,31 ----
  #include "timeoutwrite.h"
  #include "commands.h"
  
+ extern char *malloc();
+ 
  #define MAXHOPS 100
  unsigned int databytes = 0;
  int timeout = 1200;
***************
*** 96,101 ****
--- 98,106 ----
  int bmfok = 0;
  stralloc bmf = {0};
  struct constmap mapbmf;
+ int flagsh = 0;
+ stralloc sh = {0};
+ struct constmap mapsh;
  
  void setup()
  {
***************
*** 117,122 ****
--- 122,132 ----
    if (bmfok)
      if (!constmap_init(&mapbmf,bmf.s,bmf.len,0)) die_nomem();
   
+   flagsh = control_readfile(&sh,"control/senderhosts",0);
+   if (flagsh == -1) die_control();;
+   if (flagsh)
+     if (!constmap_init(&mapsh,sh.s,sh.len,0)) die_nomem();
+ 
    if (control_readint(&databytes,"control/databytes") == -1) die_control();
    x = env_get("DATABYTES");
    if (x) { scan_ulong(x,&u); databytes = u; }
***************
*** 208,213 ****
--- 218,249 ----
    return 0;
  }
  
+ static stralloc host = {0};
+ 
+ int senderhosts(buf,len)
+ char *buf;
+ int len;
+ {
+   int j;
+ 
+   if (!flagsh) return 1;
+ 
+   j = byte_rchr(buf,len,'@');
+   if (j >= len) return 0; /* null domain in relaying is not acceptable */
+ 
+   ++j; buf += j; len -= j;
+ 
+   if (!stralloc_copyb(&host,buf,len)) return -1;
+   buf = host.s;
+   case_lowerb(buf,len);
+ 
+   for (j = 0;j < len;++j)
+     if (!j || (buf[j] == '.'))
+       if (constmap(&mapsh,buf + j,len - j)) return 1;
+ 
+   return 0;
+ }
+ 
  int addrallowed()
  {
    int r;
***************
*** 241,246 ****
--- 277,286 ----
  {
    if (!addrparse(arg)) { err_syntax(); return; }
    flagbarf = bmfcheck();
+   if (senderhosts(addr.s,str_len(addr.s))) {
+     relayclient = malloc(sizeof(char *));
+     strcpy(relayclient,"\0");
+   }
    seenmail = 1;
    if (!stralloc_copys(&rcptto,"")) die_nomem();
    if (!stralloc_copys(&mailfrom,addr.s)) die_nomem();
