--- mh.c.orig	Thu Sep 20 23:23:32 2001
+++ mh.c	Mon Sep 24 12:01:30 2001
@@ -1218,6 +1218,7 @@
   struct maildir **last;
   HASH *fnames;
   int i, j, deleted;
+  FILE *fp;
   
   if(!option (OPTCHECKNEW))
     return 0;
@@ -1230,8 +1231,24 @@
     
     snprintf (buf, sizeof (buf), "%s/.mh_sequences", ctx->path);
     if (stat (buf, &st_cur) == -1)
-      modified = 1;
-    
+    {
+      st_cur.st_mtime = (time_t)0;
+      if (!ctx->readonly)
+      {
+	/* generate .mh_sequences file */
+	if ((fp = safe_fopen (buf, "w")) != NULL)
+	{
+	  safe_fclose (&fp);
+	  stat (buf, &st_cur);
+	  ctx->mtime_cur = st_cur.st_mtime;
+	  strfcpy (buf, ctx->path, sizeof(buf));
+	  stat (buf, &st);
+	  ctx->mtime = st.st_mtime;
+	}
+	else
+	  ctx->readonly = 1;
+      }
+    }
   }
   else if(ctx->magic == M_MAILDIR)
   {
