diff -uNr Text-MeCab-0.12.orig/Build.PL Text-MeCab-0.12/Build.PL
--- Text-MeCab-0.12.orig/Build.PL	2006-07-15 12:36:27.000000000 +0900
+++ Text-MeCab-0.12/Build.PL	1970-01-01 09:00:00.000000000 +0900
@@ -1,41 +0,0 @@
-#!perl
-# $Id: /mirror/Text-MeCab/trunk/Build.PL 2079 2006-07-15T03:24:27.238091Z daisuke  $
-#
-# Copyright (c) 2006 Daisuke Maki <dmaki@cpan.org>
-# All rights reserved.
-
-use strict;
-use File::Spec;
-use Module::Build;
-
-my $result = do 'tools/probe_mecab.pl';
-die if $@;
-
-my $class = Module::Build->subclass(code => <<'CODE');
-    sub ACTION_disttest
-    {
-        my $self = shift;
-        local $ENV{PERL_TESTPOD} = 1;
-        $self->SUPER::ACTION_disttest(@_);
-    }
-
-    sub find_test_files
-    {
-        my $self  = shift;
-        my $tests = $self->SUPER::find_test_files(@_);
-
-        return $tests if $ENV{PERL_TESTPOD};
-        return [ grep { $_ !~ /\bpod.*\.t\z/ } @$tests ];
-    }
-CODE
-
-my $build = $class->new(
-    dist_name => 'Text-MeCab',
-    dist_version_from => 'lib/Text/MeCab.pm',
-    extra_compiler_flags => [ split(/\s+/, $result->{cflags}), '-Wall' ],
-    extra_linker_flags  => [ split(/\s+/, $result->{libs}) ],
-    c_source => 'src',
-    license => 'perl',
-    create_readme => 1,
-);
-$build->create_build_script;
\ No newline at end of file
diff -uNr Text-MeCab-0.12.orig/MANIFEST Text-MeCab-0.12/MANIFEST
--- Text-MeCab-0.12.orig/MANIFEST	2006-07-15 12:36:27.000000000 +0900
+++ Text-MeCab-0.12/MANIFEST	2006-07-29 00:24:21.000000000 +0900
@@ -1,4 +1,3 @@
-Build.PL
 Changes
 lib/Text/MeCab.pm
 lib/Text/MeCab.xs
diff -uNr Text-MeCab-0.12.orig/Makefile.PL Text-MeCab-0.12/Makefile.PL
--- Text-MeCab-0.12.orig/Makefile.PL	2006-07-15 12:36:27.000000000 +0900
+++ Text-MeCab-0.12/Makefile.PL	2006-07-29 00:24:38.000000000 +0900
@@ -5,7 +5,7 @@
 # All rights reserved.
 
 use strict;
-use ExtUtils::MakeMaker 6.25; # Avoid Build.PL bug
+use ExtUtils::MakeMaker;
 
 my $result = do 'tools/probe_mecab.pl';
 die if $@;
