diff -ur Text-MeCab-0.13.orig/tools/probe_mecab.pl Text-MeCab-0.13/tools/probe_mecab.pl
--- Text-MeCab-0.13.orig/tools/probe_mecab.pl	2006-08-08 01:13:13.000000000 +0900
+++ Text-MeCab-0.13/tools/probe_mecab.pl	2006-09-11 00:29:54.000000000 +0900
@@ -102,7 +102,7 @@
     print "No linker flags specified\n";
 }
 
-my $encoding = 'euc-jp';
+my $encoding = 'utf8';
 print 
     "Text::MeCab needs to know what encoding you built your dictionary with\n",
     "to properly execute tests.\n",
